Docs: Clarify requirements for up-to-date-ness of bucket index (#4314)
grafana/mimir: docs/sources/mimir/operators-guide/architecture/bucket-index/index.md

## Benefits

> The [querier]({{< relref "../components/querier.md" >}}), [store-gateway]({{< relref "../components/store-gateway.md" >}}) and [ruler]({{< relref "../components/ruler/index.md" >}}) must have an almost[^1] up-to-date view of the storage bucket, in order to find the right blocks to lookup at query time (querier) and load block's [index-header]({{< relref "../binary-index-header.md" >}}) (store-gateway).

Reviewer: ```suggestion
The [querier]({{< relref "../components/querier.md" >}}), [store-gateway]({{< relref "../components/store-gateway.md" >}}) and [ruler]({{< relref "../components/ruler/index.md" >}}) must have an almost[^1] up-to-date view of the storage bucket, in order to find the right blocks to look up at query time (querier) and load block's [index-header]({{< relref "../binary-index-header.md" >}}) (store-gateway).
```

Reviewer: "a" block, "the" block, or plural?

Author: fixed in 153f59713

## How it's used by the store-gateway

The [store-gateway]({{< relref "../components/store-gateway.md" >}}), at startup and periodically, fetches the bucket index for each tenant that belongs to its shard, and uses it as the source of truth for the blocks and deletion marks in the storage. This removes the need to periodically scan the bucket to discover blocks belonging to its shard.

> [^1]: Ingesters regularly add new blocks to the bucket as they offload data to long-term storage, and compactors subsequently compact these blocks and mark the original blocks for deletion. Actual deletion happens after a delay (`compactor.deletion-delay`). Trying to fetch a block that has already been deleted will lead to query failures. Therefore, in this context, an _almost up-to-date_ view is a view that's outdated by less than `compactor.deletion-delay`.

Reviewer 2: ```suggestion
[^1]: Ingesters regularly add new blocks to the bucket as they offload data to long-term storage, and compactors subsequently compact these blocks and mark the original blocks for deletion. Actual deletion happens after a delay (`-compactor.deletion-delay`). Trying to fetch a block that has already been deleted will lead to query failures. Therefore, in this context, an _almost up-to-date_ view is a view that's outdated by less than `-compactor.deletion-delay`.
```

Author: > in this context

is meant to refer to the use of the bucket index by `querier`s, `store-gateway`s and `ruler`s (that's what the sentence in the doc is referring to). Do these components vary in their use of the bucket index with respect to how much out of date the bucket index can be without causing failures?

Author: > I find it a bit difficult precisely describe what almost up-to-date means because it depends on the actual context.

If you find it difficult to describe precisely I think that's even more reason to pick the different meanings apart and describe them well for each context they're used in.

> [^1]: Ingesters regularly add new blocks to the bucket as they offload data to long-term storage, and compactors subsequently compact these blocks and mark the original blocks for deletion. Actual deletion happens after a delay (`compactor.deletion-delay`). Trying to fetch a block that has already been deleted will lead to query failures. Therefore, in this context, an _almost up-to-date_ view is a view that's outdated by less than `compactor.deletion-delay`.

Reviewer: ```suggestion
[^1]: Ingesters regularly add new blocks to the bucket as they offload data to long-term storage,
and compactors subsequently compact these blocks and mark the original blocks for deletion.
Actual deletion happens after the delay value that is associated with the parameter `compactor.deletion-delay`.
If you try to fetch a block that has already been deleted, doing so will lead to query failures.
Therefore, in this context, an _almost up-to-date_ view is a view that’s outdated by less than the value of `compactor.deletion-delay`.
```

Reviewer: * Broke up lines based on meaning;
* introduced parameter and parameter value;
* shifted passive to active voice in "If you try to fetch a block..." <-- is a human doing this, ultimately?

Reviewer: How about, "If a deleted block is fetched, query failures occur."?

Author: technically, a deleted block cannot be fetched (it's gone), how about "an attempt to fetch a deleted block will lead to failure of the query"?
